src/uopool/mod.rs

```diff
@@ -1,36 +1,52 @@
 use crate::{
     types::user_operation::{UserOperation, UserOperationHash},
-    uopool::{server::uopool_server::uo_pool_server::UoPoolServer, services::UoPoolService},
+    uopool::{
+        memory::MemoryMempool, server::uopool_server::uo_pool_server::UoPoolServer,
+        services::UoPoolService,
+    },
 };
 use anyhow::Result;
+use async_trait::async_trait;
 use clap::Parser;
 use educe::Educe;
-use ethers::types::Address;
+use ethers::{
+    abi::AbiEncode,
+    types::{Address, H256, U256},
+    utils::keccak256,
+};
 use jsonrpsee::tracing::info;
 use parking_lot::RwLock;
-use std::{collections::HashMap, net::SocketAddr, sync::Arc, time::Duration};
+use std::{collections::HashMap, fmt::Debug, net::SocketAddr, sync::Arc, time::Duration};
 
+pub mod memory;
 pub mod server;
 pub mod services;
 
-#[derive(Educe)]
-#[educe(Debug)]
-pub struct UserOperationPool {
-    pub pool: Arc<RwLock<HashMap<UserOperationHash, UserOperation>>>,
+pub type MempoolId = H256;
+
+pub fn mempool_id(entry_point: Address, chain_id: U256) -> MempoolId {
+    H256::from_slice(keccak256([entry_point.encode(), chain_id.encode()].concat()).as_slice())
 }
 
-impl UserOperationPool {
-    pub fn new() -> Self {
-        Self {
-            pool: Default::default(),
-        }
-    }
+#[async_trait]
+pub trait Mempool: Debug + Send + Sync + 'static {
+    async fn add(
+        &mut self,
+        user_operation: UserOperation,
+        entry_point: Address,
+        chain_id: U256,
+    ) -> anyhow::Result<UserOperationHash>;
+    async fn get(&self, user_operation_hash: UserOperationHash) -> anyhow::Result<UserOperation>;
+    async fn get_all(&self) -> anyhow::Result<Vec<UserOperation>>;
+    async fn get_all_by_sender(&self, sender: Address) -> anyhow::Result<Vec<UserOperation>>;
```

Comment:
Just a nitpick. I think we could also turn the `get_all` and `get_all_by_sender` into an `Iterator` trait method which could simplify the trait even more.